Feature: BB-167 inject oplog for specific buckets

welansari opened the pull request:
Issue: [BB-167](https://scality.atlassian.net/browse/BB-167)

**Changes:**

NotificationConfigManager:
- Mongo changeStream delete events don't have the `fullDocument` field, which throws an error when trying to get the object id from it. Switched to the `documentKey` field instead as it is always provided

OplogPopulator:
- Now configures Kafka-connect mongo connector to only listen to specific buckets (collections) that have at least one backbeat extension active.
- A Mongo change stream is established to listen to `__metastore` changes and reconfigure the Kafka-connect connector accordingly.

ChangeStreamWrapper:
- Created simple class to handle change streams

KafkaConnectWrapper:
- Changed location of file

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -64,10 +63,61 @@ class OplogPopulator {
+            const client = await this._MongoClientConnect(this._mongoUrl, {
+                 replicaSet: this._replicaSet,
+                 useNewUrlParser: true,
+            });
```
can't we should pass `params.mongoConfig` ?
the param in backbeat config is the same format used by MongoClient, so this is usually enough I think

welansari replied:
We are not using arsenal's MongoClient here so params need to be adjusted
we are using the MongoClient provided by the mongodb library, this is done because we won't be using any of the arsenal implementation functions (not needed)

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -64,10 +63,61 @@ class OplogPopulator {
+     * @param {string} connectorConfig.name connector name
+     * @returns {Object} connector configuration
+     */
+    _getDefaultConnectorConfiguration(connectorConfig) {
```
is it a getter (as the name of the function) or a setter (as in the comment) ?
seems awfully similar to some mongodb connection code, also; don't we have it somewhere already, where we can use it?

welansari replied:
~~The mongodb connection code you are referring to i'm assuming is in arsenal's MongoClientInterface ? this part of the code is directly in the constructor so it can't be used~~
Fixed

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -77,30 +127,262 @@ class OplogPopulator {
+        };
+        // setup filters
+        this._activeExtensions.forEach(extName => {
+            const field = constants.extensionNameToBucketMetadataConfigField[extName];
```
`extensionNameToBucketMetadataConfigField` : may want to try and find a better name...

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -77,30 +127,262 @@ class OplogPopulator {
+
+    /**
+     * Handler for the change stream "change" event.
+     * Updated connector pipeline when change occurs
```
```suggestion
     * Updates connector pipeline when change occurs
```

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -77,30 +127,262 @@ class OplogPopulator {
+     */
+    _handleChangeStreamChangeEvent(change) {
+        // caching change stream resume token
+        this._changeStreamResumeToken = change._id;
```
I think there was some (generic) code from Vianney to implement this "scan & watch" pattern : can't we use it?

welansari replied:
the code is not yet merged, so it cannot be used
Will use a simple wrapper class for now, for better code readability

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-167 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.1`


Please check the `Fix Version/s` of BB-167, or the target
branch of this pull request.

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -77,30 +127,262 @@ class OplogPopulator {
+        // updating list of buckets to listen to
+        const idx = this._bucketsToListenTo.indexOf(change.documentKey._id);
+        // no fullDocument field in delete events
+        const isBackbeatEnabled = change.fullDocument ? this._isBucketBackbeatEnabled(change.fullDocument.value)
```
instead of receiving all events, we may be able to add the same filter we use to "read" the current list (e.g. `_getBackbeatEnabledBuckets`) direclty in the change stream ?

welansari replied:
This will cause an issue if a bucket becomes invalid (no extension active), as the change stream event that invalidates it will be ignored and so the connector pipeline won't be updated

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -77,30 +127,262 @@ class OplogPopulator {
+        // caching change stream resume token
+        this._changeStreamResumeToken = change._id;
+        // updating list of buckets to listen to
+        const idx = this._bucketsToListenTo.indexOf(change.documentKey._id);
```
searching in an array should not be done : unless we can be sure the array size will always be small (not the case here)

may be acceptable for an initial implementation (we know things will change) ; but if there is a chance this code stays (e.g. we may still need this kind of lookup, even in next step(s) of implementation) : then we need to come up with a more efficient approach.

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -77,30 +127,262 @@ class OplogPopulator {
+            });
+            throw errors.InternalError.customizeDescription(err.message);
+        }
+    }
```
looking at the code here, I think we may benefit from extracting a generic "change stream" wrapper class : e.g. to handle errors, resume token, ... (which is done here, and in a few other places)

(unless we can use/extend existing implementation of the scan&watch pattern)

francoisferrand commented on extensions/oplogPopulator/OplogPopulatorTask.js:
```diff
@@ -1,34 +1,53 @@
 (async () => {
     try {
         await oplogPopulator.setup();
+        schedule.scheduleJob(qpConfig.cronRule, async () => {
```
why do we need to do this periodically?
update is required only :
- on startup
- when the list of buckets to watch has changed

francoisferrand commented on extensions/oplogPopulator/OplogPopulatorTask.js:
```diff
@@ -1,34 +1,53 @@
+        schedule.scheduleJob(qpConfig.cronRule, async () => {
+            try {
+                // Update each connector's config
+                await Promise.all(oplogPopulatorConfig.connectors.map(connConf =>
```
is this tacking more than BB-167 ? i think we should have a single connector for now...
dispatching onto multiple connectors is next ticket (e.g. BB-168) ...

francoisferrand commented on extensions/oplogPopulator/constants.js:
```diff
@@ -5,7 +5,13 @@ const constants = {
-    }
+    },
+    extensionNameToBucketMetadataConfigField: {
+        notification: 'notificationConfiguration',
```
is it enough to check the presence of these field in the bucket metadata?
e.g. can we have the field with `enabled=false` inside?

francoisferrand reviewed (commented):
some cleanup to do, and I am really concerned about the "multiple connectors" code

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-167 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.2`


Please check the `Fix Version/s` of BB-167, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

WillToozsScality reviewed (approved):
for what it's worth, seems ok to me

jonathan-gramain commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -27,80 +34,390 @@ class OplogPopulator {
+        this._database = this._mongoConfig.database;
+
+        // Promisifying callback functions
+        this._MongoClientConnect = util.promisify(MongoClient.connect);
```
is there a reason to capitalize `Mongo` instead of `_mongoClientConnect`

welansari replied:
Forgot `Mongo` supported async 😅, fixed it

jonathan-gramain commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -27,80 +34,390 @@ class OplogPopulator {
+        if (this._activeExtensions.includes('lifecycle')) {
+            const field = constants.extensionConfigField.lifecycle;
+            const extFilter = {};
+            // processing buckets where at least one replication
```
replication -> lifecycle

jonathan-gramain commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -27,80 +34,390 @@ class OplogPopulator {
+                .toArray();
+            return buckets;
+        } catch (err) {
+            this._logger.error('Could not connect to MongoDB', {
```
Maybe a more generic message like `error querying MongoDB` could be more appropriate, in case the error is not about connectivity

jonathan-gramain commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -27,80 +34,390 @@ class OplogPopulator {
+            return buckets;
+        } catch (err) {
+            this._logger.error('Could not connect to MongoDB', {
+                method: 'OplogPopulator._setupMongoClient',
```
`OplogPopulator._getBackbeatEnabledBuckets`

jonathan-gramain commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -27,80 +34,390 @@ class OplogPopulator {
+            if (!rules || rules.length === 0) {
+                areExtensionsEnabled.push(false);
+            } else {
+                const areRulesActive = bucketMetadata[field].rules.some(rule =>
```
```
const areRulesActive = rules.some(rule =>
```

jonathan-gramain commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -27,80 +34,390 @@ class OplogPopulator {
+                break;
+        }
+        this._logger.debug('Change stream event processed', {
+            method: 'OplogPopulator._handleChangeStreamChange',
```
could add

```
type: change.operationType
```

jonathan-gramain commented on tests/unit/lib/wrappers/ChangeStreamWrapper.js:
```diff
@@ -0,0 +1,200 @@
+            .catch(err => assert.ifError(err));
+        });
+
+        it('should close then reset the change steam on error', async () => {
```
steam -> stream

jonathan-gramain reviewed (commented):
I did a partial review only.

This PR is too large, it probably explains why it's hard to get timely reviews from NA ;)

Try to see if you can split it into smaller meaningful chunks, or for the next ones if it's too hard or late for this one.

francoisferrand commented on lib/wrappers/ChangeStreamWrapper.js:
```diff
@@ -0,0 +1,95 @@
+ *
+ * @classdesc Wrapper arround MongoDB change stream API
+ */
+class ChangeStreamWrapper {
```
just `MongoChangeStream` would be enough?
The "Wrapper" suffix does not add any informatin...

Commit:
feature: BB-167 move KafkaConnectWrapper and fix delete response not having body

Change location of KafkaConnectWrapper class

Fixed issue when deleting a connector where the response
does not have a body

Commit:
feature: BB-167 Add change stream wrapper class

Added a change stream class that listen and retreive changes made
to a MongoDb collection

Commit:
feature: BB-167 add mongo helper function that constructs connection string

Commit:
feature: BB-167 update oplogPopulator config

Commit:
feature: BB-167 update OplogPopulator task

OplogPopulator now needs to know the active extensions to correctly
configure the kafka-connect mongo connector to listen to buckets that
use these extensions

Commit:
feature: BB-167 add dynamic kafka-connect connector configuration

Changed the oplog populator to dynamically configure the kafka-connect mongo
connector to listen to specific buckets that use at least one of the active
backbeat extensions.

All buckets that use at least one extension are first listed and then added to
the configuration of the kafka-connect mongo connector (only one connector is
used for now).

A change stream is then established to listen to changes on the __metastore
collection. The connector is then dynamically reconfigured to start or stop
listening to buckets that have been added, removed or updated in the collection.

Commit:
feature: BB-167 fix invalid id in change stream delete events

Mongo changeStream delete events don't have the fullDocument field,
which throws an error when trying to get the object id from it.

Switched to the documentKey field instead as it is always provided

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -1,10 +1,15 @@
 const joi = require('joi');
+const { errors } = require('arsenal');
+const MongoClient = require('mongodb').MongoClient;
```
nit: couldn't we do `const { MongoClient } = require('mongodb')` instead?

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-167 contains:


* `8.4.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.3`


Please check the `Fix Version/s` of BB-167, or the target
branch of this pull request.

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -27,80 +32,388 @@ class OplogPopulator {
+        if (this._activeExtensions.includes('notification')) {
+            const field = constants.extensionConfigField.notification;
+            const extFilter = {};
+            // notification config should be an object
+            extFilter[`value.${field}`] = { $type: 3 };
+            filter.$or.push(extFilter);
+        }
+        // replication filter
+        if (this._activeExtensions.includes('replication')) {
+            const field = constants.extensionConfigField.replication;
+            const extFilter = {};
+            // processing buckets where at least one replication
+            // rule is enabled
+            extFilter[`value.${field}.rules`] = {
+                $elemMatch: {
+                    enabled: true,
+                },
+            };
+            filter.$or.push(extFilter);
+        }
+        // lifecycle filter
+        if (this._activeExtensions.includes('lifecycle')) {
+            const field = constants.extensionConfigField.lifecycle;
+            const extFilter = {};
+            // processing buckets where at least one lifecycle
+            // rule is enabled
+            extFilter[`value.${field}.rules`] = {
+                $elemMatch: {
+                    ruleStatus: 'Enabled',
+                },
+            };
+            filter.$or.push(extFilter);
+        }
+        // ingestion filter
+        if (this._activeExtensions.includes('ingestion')) {
+            const field = constants.extensionConfigField.ingestion;
+            const extFilter = {};
+            extFilter[`value.${field}.status`] = 'enabled';
+            filter.$or.push(extFilter);
+        }
```
ideally, I would prefer this logic (e.g. the mapping from "extension" to associated filter) to be outside of this code : possibly be part of each extension itself.

So that we don't need to modify oplogPopulator itself, and simply need to add the required "support" in the extension.

(to be clear: the part which should not be here IMO is the "business logic" of each extension, i.e. knowing that _for ingestion we need whatever field to be set to enabled_. There may however be code here to loop over "something" and generically call that to build the actual mongo filter)

welansari replied:
Techdebt ticket created : [BB-194](https://scality.atlassian.net/browse/BB-194)

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -27,80 +32,388 @@ class OplogPopulator {
+        }
+        // replication extension
+        if (this._activeExtensions.includes('replication')) {
+            const field = constants.extensionConfigField.replication;
```
same as above, the business logic should be moved out of this (generic) code, somehow ⬆️

Note that this may be done in a separate techdebt ticket, to keep momentum and avoid conflicts/merges (→ please create it, in that case)

welansari replied:
Techdebt ticket created

francoisferrand commented on extensions/oplogPopulator/OplogPopulatorTask.js:
```diff
@@ -10,10 +10,12 @@ werelogs.configure({ level: config.log.logLevel,
 
 const mongoConfig = config.queuePopulator.mongo;
 const oplogPopulatorConfig = config.extensions.oplogPopulator;
+const activeExtensions = Object.keys(config.extensions);
```
not correct : should only include the extensions which are actually using oplogPopulator...

e.g. nothing for now :-/ ....though we may enable it for bucket notification already anyway, since we are transitioning it and it is not yet used in 8.x. This will make testing easier, with no product impact in the meantime.

francoisferrand replied:
In addition, maybe we may need to check a bit deeper?
I am thinking of the case (which may not be possible, to check) where the config contains some details about an extension, but it is not fully enabled (e.g. something like `config = { "extensions": [ "replication": { "enabled": false } ] }`)

francoisferrand reviewed (changes requested):
oplog populator should not be enabled for all "active" extensions yet : only for these which require it (e.g. nothing yet, or possibly bucket-notification to make ongoing dev easier)

welansari commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-167 contains:


* `8.4.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.3`


Please check the `Fix Version/s` of BB-167, or the target
branch of this pull request.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`
* `stabilization/7.10.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`
* `stabilization/7.10.4`



Please check the status of the associated issue BB-167.

Goodbye kerkesni.
